java/runtime/src/main/java/io/ray/runtime/object/ObjectRefImpl.java

```diff
@@ -53,4 +47,46 @@ public ObjectId getId() {
     return type;
   }
 
+  @Override
+  protected void finalize() throws Throwable {
```

Comment:
I think there would be many `object ref`, using `finalize` is definitly not a recommanded way. It will impact gc performance a lot. How about using `phantom reference`

Reply:
Another question is since java gc is not eager, how the system beheave before gc and after there is no reference object of the `ObjectRefImpl`. There can be many ghost reference object which can't be referenced and still not be gc.

Reply:
Hmm so is there a way we can guarantee that once java gc happens, `ObjectRef`s that aren't in scope anymore will get deleted?

There's a similar problem in Python where the GC doesn't collect some refs due to cycles, and the `ObjectRef` itself is small so having a lot of unreachable `ObjectRef`s doesn't create enough memory pressure to trigger a full GC sweep. So when we hit an OOM, we trigger a full GC sweep on all Python workers to try and collect any `ObjectRef`s ([here](https://github.com/ray-project/ray/blob/ce3f5427394af288f9699bed3d479c47181b252f/python/ray/_raylet.pyx#L581) is the GC callback that we give to the C++ worker to call). Is there an equivalent in Java that we should add?

Reply:
If `ObjectRefs` lives long enough, it will locate in old generation of java heap. Since the object refs are small objects, having many object refs may not trigger java gc. Even it does, java gs most happen in new generation of java heap, which will still miss many objects refs in old generation, thus some non-reachable objects refs can't still be not collected. When there is no enough memory in object store, even we call `System.gc`, it's just a hint to java gc, and can't promise that java scan old generation to put those object refs to reference queue so that we can be notified eventually.

Reply:
Hmm I see, thanks for the info! I think it probably still makes sense to implement the Java callback to `System.gc`, though, right? Maybe it won't be able to clean up object refs in the old generation, but it seems better than doing nothing.

Reply:
Well, I ended up using weak references instead of phantom references. According to https://web.archive.org/web/20110303225354/http://weblogs.java.net/blog/2006/05/04/understanding-weak-references, I think there's no additional benefit in using phantom references. Besides, weak references are enqueued before finalization while phantom references are enqueued after. So we can remove local references of object refs earlier.

As @chaokunyang said, GC in Java is hard to control. But let's try this first. I can't think of a better approach yet. 😅

I've added the callback to invoke `System.gc`, but I haven't tested it. @stephanie-wang Do you know how to trigger the `gc_collect` callback?

Reply:
Hmm maybe you can try mimicking these tests from Python? https://github.com/ray-project/ray/blob/master/python/ray/tests/test_global_gc.py

I think those tests are actually a bit flaky in the CI, but you could running locally for now.

Reply:
I've added a test case for global GC in Java.